mlflow/dspy/save.py

```diff
@@ -0,0 +1,294 @@
+    dspy_settings = dict(dspy.settings.config)
+    if "trace" in dspy_settings:
+        # Don't save the trace in the model, which is only useful during the training phase.
+        del dspy_settings["trace"]
+
+    # Store both dspy model and settings in `DspyModelWrapper` for serialization.
+    wrapped_dspy_model = DspyModelWrapper(model, dspy_settings)
+
+    with open(model_path, "wb") as f:
+        cloudpickle.dump(wrapped_dspy_model, f)
+
+    flavor_options = {
+        "model_path": model_subpath,
+        "dspy_version": version("dspy-ai"),
+    }
+
+    # Add flavor info to `mlflow_model`.
+    mlflow_model.add_flavor(FLAVOR_NAME, **flavor_options)
+
+    # Add loader_module, data and env data to `mlflow_model`.
+    pyfunc.add_to_model(
+        mlflow_model,
+        loader_module="mlflow.dspy",
+        conda_env=_CONDA_ENV_FILE_NAME,
+        python_env=_PYTHON_ENV_FILE_NAME,
+    )
+
+    # Add model file size to `mlflow_model`.
+    if size := get_total_file_size(path):
+        mlflow_model.model_size_bytes = size
+
+    # Save mlflow_model to path/MLmodel.
+    mlflow_model.save(os.path.join(path, MLMODEL_FILE_NAME))
+
+    if conda_env is None:
+        if pip_requirements is None:
+            default_reqs = get_default_pip_requirements()
+            # To ensure `_load_pyfunc` can successfully load the model during the dependency
+            # inference, `mlflow_model.save` must be called beforehand to save an MLmodel file.
+            inferred_reqs = infer_pip_requirements(path, FLAVOR_NAME, fallback=default_reqs)
+            default_reqs = sorted(set(inferred_reqs).union(default_reqs))
+        else:
+            default_reqs = None
+        conda_env, pip_requirements, pip_constraints = _process_pip_requirements(
+            default_reqs,
+            pip_requirements,
+            extra_pip_requirements,
+        )
+    else:
+        conda_env, pip_requirements, pip_constraints = _process_conda_env(conda_env)
+
+    # dspy's pypi name is dspy-ai, so we need to remove "dspy" from the
+    # auto-generated pip_requirements.
+    pip_requirements = list(filter(lambda x: not x.startswith("dspy=="), pip_requirements))
+    conda_env["dependencies"][-1]["pip"] = pip_requirements
+    with open(os.path.join(path, _CONDA_ENV_FILE_NAME), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    # Save `constraints.txt` if necessary.
+    if pip_constraints:
+        write_to(os.path.join(path, _CONSTRAINTS_FILE_NAME), "\n".join(pip_constraints))
+
+    # Save `requirements.txt`.
+    write_to(os.path.join(path, _REQUIREMENTS_FILE_NAME), "\n".join(pip_requirements))
+
+    _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))
+
+
+@experimental
+@format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
+def log_model(
+    model,
+    artifact_path,
+    conda_env=None,
+    signature: ModelSignature = None,
+    input_example: ModelInputExample = None,
```

Comment:
`model_config`, `**kwargs`, and `code_paths` should be added.

Reply:
done